discuss.1
.\"
.\"
.\"
.\"
.\"
.\"
.TH DISCUSS 1 "14 April 1987" "debathena-discuss" "MIT SIPB"
.SH NAME
discuss \- Conferencing system
.SH SYNOPSIS
.B discuss
[ 
.B meeting_name
]
.SH DESCRIPTION
.I Discuss
is a user-interface program to a networked conferencing system.  It
provides the concept of "meetings", which a user may attend.  Users
may post "transactions" (short pieces of text) which may be read out
of the meeting.  The
.I Discuss
user interface is heavily influenced by the Multics
.I Forum
conferencing system.

The commands accepted may be listed by using the "?" command once in
discuss.  There is also a "help" command which provides help on a
command once inside the program.  
Type "help commandname" to get information on using a command.
The "quit" command exits discuss.

For an introduction to using this system, consult the document "Using
Discuss", available from the SIPB.

.SH FILES
.ta \w'/var/spool/discuss/*  'u
.br
/var/spool/discuss/*	Meetings
.br
/tmp/mtg#####.#####	Temporary file for editing.
.br
~/.meetings	Data file listing meetings that you attend.

.SH AUTHORS
Stan Zanarotti, Ken Raeburn, Robert French, Bill Sommerfeld, and Ted
Ts'o of the MIT Student Information Processing Board.

.SH "SEE ALSO"
.I "Using Discuss",
available from the SIPB.

.I Multics forum manual
:\-)

.SH BUGS
Many; Let us know about them.  Many of them are documented in the
online helpfiles.






debian/changelog
debathena-discuss (10.0.15-0debathena2) unstable; urgency=low

  * Bump debian/compat to 6
  * Package discuss server utilities (Trac: #881)
  * Cleanup man pages so they don't generate errors
  * Add missing man pages to satisfy Lintian
  * Comply with Lintian's feelings on grammar
  * Fix buffer overflow when re-adding meeting (Trac: #973)

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 13 Sep 2012 17:02:56 -0400

debathena-discuss (10.0.15-0debathena1) unstable; urgency=low

  * Fix FTBFS with -Werror=format-security.

 -- Geoffrey Thomas <geofft@mit.edu>  Sat, 28 Apr 2012 20:36:38 -0700

debathena-discuss (10.0.14-0debathena1) unstable; urgency=low

  * edsc now defines and sets its own com_err hook to avoid libdsk
    confusing emacs (Trac: 603)

 -- Jonathan Reed <jdreed@mit.edu>  Sun, 25 Sep 2011 17:07:38 -0400

debathena-discuss (10.0.13-0debathena2) unstable; urgency=low

  * Flesh out mkds.8 and rmds.8 some more.

 -- Benjamin Kaduk <kaduk@mit.edu>  Sun, 03 Jul 2011 18:31:43 -0400

debathena-discuss (10.0.13-0debathena1) unstable; urgency=low

  * Update our autogoo for the 21st century.
  * Add a working distclean target.
  * Update packaging to not require daconfiscate.
  * Eliminate libds, as its only use could be replaced by libdsk.
  * Check $PAGER for a pager before falling back on a compiled-in default
    ("more" by default, but "pager" under our packaging). (Trac: #629)

 -- Evan Broder <broder@mit.edu>  Sun, 27 Jun 2010 13:02:34 -0400

debathena-discuss (10.0.12-0debathena3) unstable; urgency=low

  * Don't depend on tcsh anymore

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 08 Apr 2010 22:41:16 -0400

debathena-discuss (10.0.12-0debathena2) unstable; urgency=low

  * Run dsc_setup on behalf of the user, like edsc does. (Trac #426)
  * Rewrite dsc_setup in bash, because all the cool kids use it.

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 08 Apr 2010 21:03:07 -0400

debathena-discuss (10.0.12-0debathena1) unstable; urgency=low

  * Stop using the obsolete AC_TYPE_SIGNAL. (Trac: #347)
  * Add missing manpages. (Trac: #266)

 -- Evan Broder <broder@mit.edu>  Sun, 15 Nov 2009 17:35:19 -0500

debathena-discuss (10.0.11-0debathena1) unstable; urgency=low

  * Move the single-letter checks in trn_expr.y:yylex to the bottom of the
    function so that they get caught after commands that start with those
    letters. (Trac: #346)

 -- Evan Broder <broder@mit.edu>  Sat, 29 Aug 2009 17:20:43 -0400

debathena-discuss (10.0.10-0debathena2) unstable; urgency=low

  * Depend on Kerberos-enabled zephyr instead of build-depending.

 -- Evan Broder <broder@mit.edu>  Mon, 27 Jul 2009 14:38:10 -0400

debathena-discuss (10.0.10-0debathena1) unstable; urgency=low

  * Patch discuss server to work without krb4. Patch by Mitch Berger.

 -- Evan Broder <broder@mit.edu>  Thu, 23 Jul 2009 21:00:55 -0400

debathena-discuss (10.0.9-0debathena1) unstable; urgency=low

  * Fix local/dumb authentication needed for mail delivery and local
    client usage. Patch by Mitch Berger. (Trac: #274)
  * Always build without krb4.

 -- Evan Broder <broder@mit.edu>  Wed, 22 Jul 2009 22:58:40 -0400

debathena-discuss (10.0.8-0debathena1) unstable; urgency=low

  * Add one-character aliases to discuss. Patch by Ben Kaduk. (Trac: #238)
  * Squash some lintian warnings by switching from ${Source-Version} to
    ${source:Version} and ${binary:Version}
  * Use HAVE_KRB4 and HAVE_KRB5 instead of KERBEROS and KERBEROS5.
  * Assorted changes to make discuss work with any combination of krb4 and
    krb5 (including neither).
  * Always build our discuss with krb5.

 -- Evan Broder <broder@mit.edu>  Mon, 18 May 2009 23:51:52 -0400

debathena-discuss (10.0.7-0debathena3) unstable; urgency=low

  * Configure the discuss inetd service.

 -- Evan Broder <broder@mit.edu>  Thu, 05 Mar 2009 15:17:20 -0500

debathena-discuss (10.0.7-0debathena2) unstable; urgency=low

  * Change DEB_AUTO_UPDATE_AUTOCONF to 2.50, not 1.
  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.
  * Clean up debian/copyright.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 14:31:41 -0500

debathena-discuss (10.0.7-0debathena1) unstable; urgency=low

  * Revert a small part of the OS X com_err patch from 10.0.6 in favor of
    changes to aclocal.m4
  * Remove hard-coded paths from discuss-mode, in favor of getting them
    from autoconf.
  * Remove debathena-machtype as a dependency, as it's no longer needed.

 -- Evan Broder <broder@mit.edu>  Fri, 09 Jan 2009 21:47:33 -0500

debathena-discuss (10.0.6-0debathena1) unstable; urgency=low

  [ Evan Broder ]
  * Add com_err support for delete and discuss on OS X.

  [ Anders Kaseorg ]
  * Depend emacs | emacsen instead of emacs22 | emacs21 | emacsen, so that
    the emacs metapackage is pulled in by default.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 08 Jan 2009 16:35:04 -0500

debathena-discuss (10.0.5-0debathena1) unstable; urgency=low

  * Fix some trivial typos.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 19 Aug 2008 16:35:07 -0400

debathena-discuss (10.0.4-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc on 64-bit platforms.
  * Fix date display in discuss on 64-bit platforms.

 -- Greg Hudson <ghudson@mit.edu>  Mon, 30 Jun 2008 12:50:56 -0400

debathena-discuss (10.0.3-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc (int/long mismatch in short_time).
  * Put edsc crash dumps in /tmp instead of /usr/tmp.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 19:39:38 -0400

debathena-discuss (10.0.2-0debathena1) unstable; urgency=low

  * Stop assuming that long is 32 bits.
  * Start assuming that int is 32 bits.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 16:29:35 -0400

debathena-discuss (10.0.1-0debathena1) unstable; urgency=low

  * Fix obsolete names for Kerberos library functions, which no longer
    worked on Mac OS X.  Patch from broder.


client/discuss.c
/*
 *
 *    Copyright (C) 1989 by the Massachusetts Institute of Technology
 *    Developed by the MIT Student Information Processing Board (SIPB).
 *    For copying information, see the file mit-copyright.h in this release.
 *
 */
/*
 *	$Id: discuss.c,v 1.60 2007-08-09 20:41:31 amb Exp $
 *
 *	A simple shell-type user interface to discuss; uses Ken Raeburn's
 *	ss library for the command interpreter.
 *
 */


#ifndef lint
static char rcsid_discuss_c[] =
    "$Id: discuss.c,v 1.60 2007-08-09 20:41:31 amb Exp $";
#endif /* lint */

#include <stdio.h>
#include <stdlib.h>
#include <sys/file.h>
#include <signal.h>
#include <string.h>
#include <sys/wait.h>
#include <pwd.h>
#include <ss/ss.h>
#include <discuss/discuss.h>
#include "config.h"
#include "globals.h"

#ifdef	lint
#define	DONT_USE(var)	var=var;
#else	/* lint */
#define	DONT_USE(var)	;
#endif	/* lint */

#define	FREE(ptr)	{ if (ptr) free(ptr); }
#define max(a, b) ((a) > (b) ? (a) : (b))

extern ss_request_table discuss_cmds;

char	dsc_version[] = "1.7";
int	sci_idx;

extern char *temp_file, *pgm, *user_id;

/* EXTERNAL ROUTINES */

tfile	unix_tfile();
char	*local_realm();
static char	buf[BUFSIZ];
char	*buffer = buf;

int main (argc, argv)
	int argc;
	char **argv;
{
	int code;
	char *initial_meeting = (char *)NULL;
	char *subsystem_name = "discuss";
	char *argv0 = argv[0];
	char *initial_request = (char *)NULL;
	bool quit = FALSE;	/* quit after processing request */
	bool flame = FALSE;	/* Have we flamed them for multiple  */

	signal(SIGPIPE, SIG_IGN);
	editor_path = getenv ("DISCUSS_EDITOR");

	while (++argv, --argc) {
		if (!strcmp(*argv, "-prompt")) {
			if (argc == 1) {
				fprintf(stderr,
					"No argument supplied with -prompt\n");
				exit(1);
			}
			argc--; argv++;
			subsystem_name = *argv;
		}
		else if (!strcmp(*argv, "-request") || !strcmp(*argv, "-rq")) {
			if (argc == 1) {
				fprintf(stderr,
					"No string supplied with -request.\n");
				exit(1);
			}
			argc--; argv++;
			initial_request = *argv;
		}
		else if (!strcmp(*argv, "-quit"))
			quit = TRUE;
		else if (!strcmp(*argv, "-no_quit"))
			quit = FALSE;
		else if (!strcmp(*argv, "-editor")) {
			if (argc == 1) {
				fprintf(stderr, "No editor name supplied with -editor\n");
				exit(1);
			}
			if (!use_editor) { 
				fprintf(stderr, "Both -editor and -no_editor specified\n");
				exit(1);
			}
			--argc;
			editor_path = *(++argv);
		}
		else if (!strcmp(*argv, "-no_editor"))
			use_editor = FALSE;
		else if (**argv == '-') {
			fprintf(stderr, "Unknown control argument %s\n",
				*argv);
			fprintf(stderr, "Usage: %s [ -prompt name ] [ -request name ] [ -quit ]\n\t\t[ -editor editor_path ] [ -no_editor ]\n",
				argv0);
			exit(1);
		}
		else {
			if (initial_meeting) {
				if (!flame) {
					fprintf(stderr, 
"More than one meeting name supplied on command line; using %s\n", 
						initial_meeting);
					flame = TRUE;
				}
			} else initial_meeting = *argv;
		}
	}

	{
		char *user;
		struct passwd *user_pw = getpwuid(getuid());
		register char *realm = local_realm();
		

		if (user_pw == NULL) {
		     fprintf(stderr,
		 "You do not appear in /etc/passwd.  Cannot continue.\n");
		     exit(1);
		}
		user = user_pw -> pw_name;
		user_id = malloc((unsigned)(strlen(user)+strlen(realm)+2));
		strcpy(user_id, user);
		strcat(user_id, "@");
		strcat(user_id, realm);
	}

	sci_idx = ss_create_invocation(subsystem_name, dsc_version,
				       (char *)NULL, &discuss_cmds, &code);
	if (code) {
	    com_err (subsystem_name, code, "creating invocation");
	    exit(1);
	}
	(void) ss_add_info_dir(sci_idx, INFO_DIR, &code);
	if (code) {
		ss_perror(sci_idx, code, INFO_DIR);
	}

#if defined(__APPLE__) && defined(__MACH__)
	add_error_table(&et_disc_error_table);
	add_error_table(&et_dsc_error_table);
#else
	initialize_disc_error_table();
	initialize_dsc_error_table();
#endif

	temp_file = malloc(64);
	pgm = malloc(64);
	(void) sprintf(temp_file, "/tmp/mtg%d.%d", (int)getuid(), getpid());

	if (code = find_rc_filename()) {
	  if (code != EACCES) {
	    fprintf(stderr, "Creating new ~/.meetings file...\n");
	    fflush(stderr);
	    char * cmd = malloc(strlen(DSC_SETUP) + strlen(" -q") + 1);

	    sprintf(cmd, "%s -q", DSC_SETUP);
	    system(cmd);
	    free(cmd);
	    code = find_rc_filename();	
	  }
	  if (code) {
	    ss_perror(sci_idx, code, "while reading meetings file.");
	    exit (1);
	  }
	}
        if (!quit) {
	     printf("Discuss version %s.  Type '?' for a list of commands.\n",
		    dsc_version);
	     if (!initial_meeting)
		  printf("\n");
	}

	if (initial_meeting != (char *)NULL) {
		(void) sprintf(buffer, "goto %s", initial_meeting);
		code = ss_execute_line(sci_idx, buffer);
		if (code != 0)
			ss_perror(sci_idx, code, initial_meeting);
	}
	if (initial_request != (char *)NULL) {
		code = ss_execute_line(sci_idx, initial_request);
		if (code != 0)
			ss_perror(sci_idx, code, initial_request);
	}
	if (!quit || code)
		code = ss_listen (sci_idx);
	leave_mtg();				/* clean up after ourselves */
	return 0;
}

int getyn(prompt,def)
    char *prompt, def;
{
	char inp[128];

	for (;;) {
		(void) printf("%s ",prompt);
		if (fgets (inp, 128, stdin) == NULL)
		    return FALSE;
		else if (inp[0] == '\n')
		    inp[0] = def;
		if (inp[0] == 'y' || inp[0] == 'Y')
		    return 1;
		else if (inp[0] == 'n' || inp[0] == 'N')
		    return 0;
		printf("Please enter 'Yes' or 'No'\n\n");
	}
}
